classes/observers/listeners/PKPUsageEventLogListener.inc.php

```diff
@@ -0,0 +1,305 @@
+            $salt = $this->_createNewSalt($saltFileName);
+            // Salt changed, flush the cache
+            $flushCache = true;
+        }
+        $currentDate = date('Ymd');
+        $saltFileLastModified = date('Ymd', filemtime($saltFileName));
+        // Create new salt if the usage log file with current date does not exist.
+        // The current usage log file will be created next (s. function handle() and logUsageEvent() above).
+        // If another process accesses this before the current usage event log file is created,
+        // consider the last modified date stamp of the salt file too.
+        if (!file_exists($logFileName) && ($currentDate != $saltFileLastModified)) {
+            $salt = $this->_createNewSalt($saltFileName);
+            // Salt changed, flush the cache
+            $flushCache = true;
+        }
+
+        if (!isset($salt)) {
+            $salt = trim(file_get_contents($saltFileName));
+        }
+
+        // Hash the IP
+        $ip = $usageEventArray['ip'];
+        $hashedIp = $this->_hashIp($ip, $salt);
+        // Never store unhashed IPs!
+        if ($hashedIp === null) {
+            assert(false);
+        }
+        $usageEventArray['ip'] = $hashedIp;
+
+        $site = Application::get()->getRequest()->getSite();
+        $context = Application::get()->getRequest()->getContext();
+        $enableGeoUsageStats = $site->getData('enableGeoUsageStats');
+        if (($enableGeoUsageStats > 0) && ($context->getData('enableGeoUsageStats') !== null) && ($context->getData('enableGeoUsageStats') < $site->getData('enableGeoUsageStats'))) {
+            $enableGeoUsageStats = $context->getData('enableGeoUsageStats');
+        }
+        // Geo data
+        $usageEventArray['country'] = $usageEventArray['region'] = $usageEventArray['city'] = null;
+        if ($enableGeoUsageStats > 0) {
+            $geoIPArray = $this->_getCachedIPLocation($ip, $hashedIp, $flushCache);
+            $usageEventArray['country'] = $geoIPArray['country'];
+            if ($enableGeoUsageStats > 1) {
+                $usageEventArray['region'] = $geoIPArray['region'];
+                if ($enableGeoUsageStats == 3) {
+                    $usageEventArray['city'] = $geoIPArray['city'];
+                }
+            }
+        }
+
+        // institutions IDs
+        $enableInstitutionUsageStats = $site->getData('enableInstitutionUsageStats');
+        if ($enableInstitutionUsageStats && ($context->getData('enableInstitutionUsageStats') !== null) && !$context->getData('enableInstitutionUsageStats')) {
+            $enableInstitutionUsageStats = $context->getData('enableInstitutionUsageStats');
+        }
+        $usageEventArray['institutionIds'] = [];
+        if ($enableInstitutionUsageStats) {
+            $institutionIds = Repo::institution()->getIdsByIP($ip, $usageEventArray['contextId'])->toArray();
+            $usageEventArray['institutionIds'] = $institutionIds;
+        }
```

Comment:
Hmmm... I am not sure if this needs to be extracted to a job -- the job would then also need to log the event once it got the institution IDs (and this would prolong the logging of the event)... 
Now I implemented it similarly to the geo caching. The following will be cached:
hashedIP => contextId => institutionIds[]
(because the salt used for the IP hashing is changed daily, the cache will be flushed daily)
Maybe this would be enough?
@asmecher, what do you think?

Reply:
If it's cached, I'm in favour of keeping the complexity low for now. If we need to offload activity from the reader's request, we can approach it later -- I do worry about relying overmuch on the jobs toolset until it's matured further.

Reply:
Yes, I implemented the caching, similar to geo data caching, and will leave it so...

I am generally afraid to use jobs for usage stats. I use them now to store the calculated metrics data into the real tables. I will need to test it VERY good once again to be sure it works as it should. Because it could lead to infinite and very difficult support cases... Maybe I should return to the scheduled tasks...